go/grpcweb/trailer.go

```diff
@@ -0,0 +1,99 @@
+//Copyright 2018 Improbable. All Rights Reserved.
+// See LICENSE for licensing terms.
+
+package grpcweb
+
+import (
+	"io"
+	"net/http"
+	"net/textproto"
+	"sort"
+	"strings"
+	"sync"
+)
+
+// gRPC-Web spec says that must use lower-case header/trailer names.
+// See "HTTP wire protocols" section in
+// https://github.com/grpc/grpc/blob/master/doc/PROTOCOL-WEB.md#protocol-differences-vs-grpc-over-http2
+type trailer struct {
+	http.Header
+}
+
+func (t trailer) Add(key, value string) {
+	lowerKey := strings.ToLower(key)
+
+	// If trailer headers, map to lower these keys.
+	if lowerKey == "trailer" {
+		value = strings.ToLower(value)
+	}
+	t.Header[lowerKey] = append(t.Header[lowerKey], value)
+}
```

Comment:
I don't think we need this function either. Lets just rely on `http.Header.Add`.

Reply:
Ok, having looked at this, how about we _keep_ this method and just remove the `Write` method (relying on `http.Header.Write`? This code is such a complicated optimisation mess that we might be able to get away with just a custom `Add`.

Reply:
Oh, just as you said, we can remove custom `Write` and rely on `http.Header.Write`!
I agree to remove custom `Write` and keep custom `Add` instead.

Reply:
I've removed it.